### PR TITLE
Fix marker rotation

### PR DESCRIPTION
**Pull request by m-kuhn**
The rotation is no longer saved in the QgsSingleSymbolRendererV2 but in the
QgsMarkerSymbolV2 in a QgsDataDefined. This commit fixes the legacy API and
redirects calls to the new API.

To check: define a rotation field on a point layer. Start editing. The "rotate point symbols" button is inactive. This commit fixes this.

@vmora Can you cross-check?

### CODE REVIEW

**Comment by m-kuhn** [comment]
And on the same page, is the following still required?
https://github.com/qgis/QGIS/blob/master/src/gui/symbology-ng/qgssinglesymbolrendererv2widget.cpp#L67-L71

**Comment by nyalldawson** [comment]
@m-kuhn your changes look like a good start to repair the API to me, but wouldn't the other renderers also need updating? I think a good improvement would be something like https://github.com/nyalldawson/QGIS/commit/27a6ecccdfbfe1b1f1a8928230d53b64b6ae57d8 (needs cleaning up), which would mean that the tool would also work for categorised/graduated renderers where only some classes have rotation fields set, and also for the rule based renderer (i vaguely remember seeing a bug report about that sometime)

**Comment by m-kuhn** [comment]
I wonder if this could be taken to the base class, graduated, single and categorized renderers use the same code in setRotationField and rotationField.

**Comment by m-kuhn** [comment]
Hmmm... The base class says this method should be implemented for renderers that support rotation. Are there some that don't?

**Comment by vmora** [comment]
@m-kuhn, sorry for the lack of reactivity (on "vacation" at the moment) 

> And on the same page, is the following still required? https://github.com/qgis/QGIS/blob/master/src/gui/symbology-ng/qgssinglesymbolrendererv2widget.cpp#L67-L71

No, this is not needed anymore.

**Comment by m-kuhn** [comment]
@vmora Thanks for the feedback anyway. Enjoy vacations.

File: src/core/symbology-ng/qgssinglesymbolrendererv2.cpp [hunk 1]
```diff
@@ -186,7 +196,7 @@ QgsFeatureRendererV2* QgsSingleSymbolRendererV2::clone() const
 {
   QgsSingleSymbolRendererV2* r = new QgsSingleSymbolRendererV2( mSymbol->clone() );
   r->setUsingSymbolLevels( usingSymbolLevels() );
-  r->setRotationField( rotationField() );
+  // r->setRotationField( rotationField() );
```
Comment (vmora) [inline src/core/symbology-ng/qgssinglesymbolrendererv2.cpp]: Is there a reason for comment .vs. simple removal ?

Comment (m-kuhn) [inline src/core/symbology-ng/qgssinglesymbolrendererv2.cpp]: That I was not sure if I was doing the right thing ;)

Comment (m-kuhn) [inline src/core/symbology-ng/qgssinglesymbolrendererv2.cpp]: I will remove your "comment-instead-of-removal" two lines down as well? ;)

**Comment by vmora** [comment]
@m-kuhn, @nyalldawson , 

> Hmmm... The base class says this method should be implemented for renderers that support rotation. Are there some that don't?

Markers support rotation, so I'm not sure why renderers should be concerned with the notion. But I understand that this would be a major API change. 

If I'm not mistaken, the setRotation/setScale is not used globally, and shouldn't, so I would go for @nyalldawson way of  tackilng the problem. 

To preserve the API, marking renderer setRotation/setScale deprecated and implementing the fix you propose seems to be a good idea. So +1 for the merge and doing the same with the other renderers.

@nyalldawson your oppinion ?

**Comment by nyalldawson** [comment]
@vmora agreed! The tricky part will be handling the rotation field getters for categorized/graduated renderers, since the field may be different or not present for all classes for the renderer. But in this case I'd say we just return an empty string.

**Comment by m-kuhn** [comment]
@nyalldawson It will not be possible to support it with this API. But if we mark it as deprecated and offer a more fine-grained access it will be fine to return an empty string for this.
Right now I'm using `mSourceSymbol` so there's no detection for different rotations.

**Comment by m-kuhn** [comment]
@vmora Is it correct that QgsDataDefined can represent a field and an expression at the same time? (E.g. field: `"a"` and expression: `"b"*5`)

Would it make sense that it does either one or the other at the same time and the method expressionString would return the field name in case it's in field-mode?

**Comment by nyalldawson** [comment]
@m-kuhn you can have both a field and an expression set, but only one is active. QgsDataDefined::useExpression will return true if the expression is being used.

**Comment by m-kuhn** [comment]
Followup: Ah, there's a `setUseExpression`. Can we do that implicitly when `setExpressionString` or `setField` is called and modify the `expressionString` accordingly?

**Comment by nyalldawson** [comment]
@m-kuhn I don't like that idea much - it means you'll no longer be able to toggle between an expression and a field in the data defined buttons without losing the expression. I do that a fair bit.

**Comment by m-kuhn** [comment]
So the latest implementation does:

 * Automatically change the mode (useExpression) when setExpressionString or setField is called
 * Adds a convenience method expressionOrField to QgsDataDefined

And adds some doxygen comments

**Comment by nyalldawson** [comment]
@m-kuhn looks good to me. How about deprecating those setRotationField/rotationField methods too? 

Also, any thoughts on the linked changes to the rotation tool? Do they look on the right track to you?

**Comment by m-kuhn** [comment]
I think it's the right track, but I don't have a complete overview of
the renderer / symbol etc. code yet.

Deprecating is fine... Will do.

**Comment by nyalldawson** [comment]
@m-kuhn @vmora #2235 is an updated version of this pr with a few extra fixes and deprecations for comment. 

I noticed that there's still some similar size field methods which are used and not deprecated, we should probably fix those too.

**Comment by m-kuhn** [comment]
@nyalldawson should we merge this PR to 2.10 and #2235 to master?

**Comment by nyalldawson** [comment]
@m-kuhn I'd say 2235 to master, then backport in a week or so to give some extra time for testing

**Comment by m-kuhn** [comment]
@nyalldawson it would be good to make a final point on this. do you have an opinion if it should be this or your code?

**Comment by nyalldawson** [comment]
@m-kuhn is there any point to backporting to 2.10 now? A new point release is very unlikely given the proximity of 2.12

**Comment by m-kuhn** [comment]
If 2.8 is unaffected... Probably no.

**Pull request closed without merging**